File: CHUONG_10/10.2_Singleton/app/src/main/java/com/example/myapplication/ui/StudentViewModel.java
package com.example.myapplication.ui;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.myapplication.data.Callback;
import com.example.myapplication.data.RemoteDataSource;
import com.example.myapplication.data.model.Student;
import com.example.myapplication.data.model.StudentList;

import java.util.List;

public class StudentViewModel extends ViewModel {
    private final MutableLiveData<List<Student>> mStudents = new MutableLiveData<>();
    private final MutableLiveData<String> mErrorMessage = new MutableLiveData<>();
    public LiveData<List<Student>> getStudents() {
        return mStudents;
    }
    public LiveData<String> getErrorMessage() {
        return mErrorMessage;
    }
    public void loadStudents() {
        RemoteDataSource dataSource = RemoteDataSource.getInstance();
        Callback<StudentList> callback = new Callback<StudentList>() {
            @Override
            public void onSuccess(StudentList data) {
                if (data != null) {
                    mStudents.postValue(data.getStudents());
                    mErrorMessage.postValue(null);
                } else {
                    mStudents.postValue(null);
                    mErrorMessage.postValue("No students data");
                }
            }

            @Override
            public void onFailure(Exception e) {
                mStudents.postValue(null);
                mErrorMessage.postValue(e.getMessage());
            }
        };
        dataSource.loadStudents(callback);
    }
}

File: CHUONG_5/5.11_RecyclerView_Kotlin/app/src/main/java/com/example/myapplication/StudentViewModel.kt
package com.example.myapplication

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider

class StudentViewModel(
    private val dataSource: LocalDataSource
): ViewModel() {
    private val _students = MutableLiveData<List<Student>>()
    public val students: LiveData<List<Student>> = _students

    fun loadStudents(){
        val studentList = dataSource.loadStudents()
        _students.value = studentList
    }

    class Factory(
        private val dataSource: LocalDataSource
    ): ViewModelProvider.Factory {
        override fun <T: ViewModel> create(modelClass: Class<T>): T{
            if (modelClass.isAssignableFrom(StudentViewModel::class.java)){
                return StudentViewModel(dataSource) as T
            }
            throw IllegalArgumentException("Unknown ViewModel class")
        }
    }
}